h2/src/main/org/h2/command/Parser.java

```diff
@@ -5021,50 +5179,66 @@ private TableView parseSingleCommonTableExpression() {
         data.session = session;
         recursiveTable = schema.createTable(data);
         session.addLocalTempTable(recursiveTable);
-        String querySQL;
-        List<Column> columnTemplateList = new ArrayList<>();
+        List<Column> columnTemplateList;
+        String[] querySQLOutput = new String[]{null};
         try {
             read("AS");
             read("(");
             Query withQuery = parseSelect();
             read(")");
-            withQuery.prepare();
-            querySQL = StringUtils.cache(withQuery.getPlanSQL());
-            ArrayList<Expression> withExpressions = withQuery.getExpressions();
-            for (int i = 0; i < withExpressions.size(); ++i) {
-                Expression columnExp = withExpressions.get(i);
-                // use the passed in column name if supplied, otherwise use alias (if used) otherwise use column name
-                // derived from column expression
-                String columnName;
-                if (cols != null){
-                    columnName = cols[i];
-                } else if (columnExp.getAlias()!=null){
-                    columnName = columnExp.getAlias();
-                }
-                else{
-                     columnName =  columnExp.getColumnName();
-                }
-                columnTemplateList.add(new Column(columnName,
-                        columnExp.getType()));
-            }
+            columnTemplateList = createQueryColumnTemplateList(cols, withQuery, querySQLOutput);
+
         } finally {
             session.removeLocalTempTable(recursiveTable);
         }
+        TableView view = createTemporarySessionView(tempViewName, querySQLOutput[0], columnTemplateList,true/*allowRecursiveQueryDetection*/, true);
+        return view;
+    }
+
+    private List<Column> createQueryColumnTemplateList(String[] cols, Query withQuery, String[] querySQLOutput) {
```

Comment:
please add comment that you are using querySQLOutput String[] as trick to get an "out" parameter

(I like to document anything "interesting" to make maintenance easier)